### DISCUSSION TITLE
Notes about improving the upgrade system

### DISCUSSION THREAD

**Discussion by marienfressinaud** [FreshRSS/FreshRSS#4251] [Strategy]
I [said](https://github.com/FreshRSS/FreshRSS/issues/4179#issuecomment-1033423488) I would share my notes about improving the upgrade system (and I almost forgot, sorry). Here we go.

**The main idea is to get rid of the server at update.freshrss.org and include the upgrade scripts directly in FRSS.** I think we can take the opportunity to improve few things as well.

Note that @Alkarex started [a PR](https://github.com/FreshRSS/FreshRSS/pull/1760) 4 years ago (:scream:), but I think we agree it's fine to restart from scratch.

**First, things we can do before changing the engine:**

- showing the installation method on the update screen (Git or archive)
- moving the installation checks on the update screen, it's the only time I think it's useful. Also, we could display warning and errors only. If there are errors, we should disable the update button
- we could include git problems (in case we detect a diff for instance)

Today, each upgrade scripts declare 5 functions:

- `apply_update`: the main function which apply the update
- `need_info_update`: return true if FRSS needs additional info to work (e.g. a new required config option)
- `ask_info_update`: to display a form to ask the additional info
- `save_info_update`: called to save the info from the form
- `do_post_update`: a function called at the end of the upgrade process, to clean things for instance (must return true)

BUT, **we almost only use `apply_update`**, I can't remember the last time I needed to write one of the other functions. I think it's fine as a first approach to don't worry with the others and simplify the system. **Also, the only thing I change between two scripts is the URL to download the package.**

Then, the main workflow is:

1. display a screen with a button to check the last version
2. ask update.freshrss.org for the update script corresponding to the current version and store two files: data/last_update.txt (containing the new version number) and data/update.php (the script itself). Example: https://update.freshrss.org/?v=1.19.1 (first line is the new version number, the rest is the script)
3. display a screen with a button to apply the update script
4. apply the script and clear the files created earlier

Obviously, things can go wrong throughout the process :)

**How can it be done differently?**

- step 2: the latest version number of FRSS should be retrieved via GitHub (https://api.github.com/repos/FreshRSS/FreshRSS/releases/latest)
- step 2: the update script (data/update.php) should be removed
- step 4: since the update script would be removed, the `apply_update` function should be moved in FRSS core (+ all the functions it depends on)
- step 4: the download url can be found in the info provided by GitHub (see `zipball_url` key), but it is redirected. I can't remember why I preferred to use the direct URL, but in case, it can be built easily: `https://codeload.github.com/FreshRSS/FreshRSS/zip/<version>`

That's basically all what have to be done… I think :) I want to add additional notes about security and privacy though (which apply to the current upgrade system as well):

- we should probably not call GitHub APIs directly, administrators may be worry about their privacy. The update.freshrss.org server could continue to exist as an optional proxy. In all cases, we should inform administrators that their servers will call an external server
- the archive hash is never checked, but it should in order to verify its integrity. I don't know if GitHub provide this information
- we could go even further by signing the releases (e.g. https://wiki.debian.org/Creating%20signed%20GitHub%20releases)

**Comment by Frenzie** [comment]
This all sounds sensible to me.
